### PR TITLE
0.10.0

### PR DESCRIPTION
**Pull request by grant-traynor**
This works with storm 0.10.0 when the storm infrastructure and dev environment (deployment environment) are running in docker containers. It needs to have certain information in the topology.yam file as the system does not pick up storm.yaml when it build the topology for submission (but does pick up defaults.yam for some reason). If would be an advantage for it to add storm.yam, then "storm" and "petrel" command lines would be equivalent.

### CODE REVIEW

File: jvmpetrel/pom.xml [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 	<groupId>storm.petrel</groupId>
 	<artifactId>storm-petrel</artifactId>
-	<version>0.9.0-SNAPSHOT</version>
```
Comment (barrywhart) [inline jvmpetrel/pom.xml]: It looks like you're removing build settings for older Storm versions. Why do this?

**Comment by barrywhart** [comment]
This PR contains two generated .jar files. Please don't include generated binary files in your PR.

File: petrel/petrel/cmdline.py [hunk 2]
```diff
@@ -42,8 +42,9 @@ def submit(sourcejar, destjar, config, venv, name, definition, logdir, extrastor
     storm_class_path = [ subprocess.check_output(["storm","classpath"]).strip(), destjar ]
     if extrastormcp is not None:
         storm_class_path = [ extrastormcp ] + storm_class_path
-    storm_home = os.path.dirname(os.path.dirname(
-        subprocess.check_output(['which', 'storm'])))
+    storm_home = os.environ['STORM_HOME']
```
Comment (barrywhart) [inline petrel/petrel/cmdline.py]: Why this change?

File: petrel/setup.py [hunk 3]
```diff
@@ -13,12 +13,12 @@
 
 PACKAGE = "petrel"
 
-PETREL_VERSION = '0.3'
+PETREL_VERSION = '0.10.0'
```
Comment (barrywhart) [inline petrel/setup.py]: Petrel version isn't supposed to match the Storm version. The Storm version makes up the last 3 parts of the overall version, e.g. 0.3.0.10.0.

File: petrel/petrel/topologybuilder.py [hunk 4]
```diff
@@ -7,14 +7,6 @@
 from petrel.generated.storm.ttypes import StreamInfo, Bolt, SpoutSpec, ShellComponent
 from petrel.generated.storm.ttypes import ComponentObject, StormTopology
 
-# Storm uses GlobalStreamId as a dict key, but the Thrift Python binding doesn't
-# support this. As a simple workaround, add a hash function that always returns
-# 0. The resulting map won't be efficient at storing large amounts of data, but
-# it should work in all cases.
-# https://issues.apache.org/jira/browse/THRIFT-162.
-assert not hasattr(GlobalStreamId, '__hash__')
-GlobalStreamId.__hash__ = lambda self: 0
```
Comment (barrywhart) [inline petrel/petrel/topologybuilder.py]: Why remove this?

File: petrel/setup.py [hunk 5]
```diff
@@ -13,12 +13,12 @@
...
 
 def get_storm_version():
     version = subprocess.check_output(['storm', 'version']).strip()
-    m = re.search(r'^(\d\.\d\.\d)(-(\w+))?', version)
-    return m.group(0), m.group(1)
+    m = re.search('^(Storm )?(\d+\.\d+\.\d+)', version)
```
Comment (barrywhart) [inline petrel/setup.py]: This looks like it will break compatibility with some older versions of Storm by removing the 

```
(-(\w+))?
```

File: jvmpetrel/pom.xml (hunk 1 above)
Comment (grant-traynor) [inline jvmpetrel/pom.xml]: OK. Not intentional. I probably reused it to create the new one.

File: petrel/petrel/cmdline.py (hunk 2 above)
Comment (grant-traynor) [inline petrel/petrel/cmdline.py]: Because in our environment the storm binary is linked, which creates a reference to /usr. STORM_HOME seems to be use elsewhere?

File: petrel/petrel/topologybuilder.py (hunk 4 above)
Comment (grant-traynor) [inline petrel/petrel/topologybuilder.py]: It caused the compilation to fail. I suspect that the bug is no longer present in the later versions of thrift?

File: petrel/setup.py (hunk 5 above)
Comment (grant-traynor) [inline petrel/setup.py]: Agree. This change was already made elsewhere in the repository.

File: petrel/setup.py [hunk 6]
```diff
@@ -59,8 +62,10 @@ def build_petrel():
     subprocess.check_call(['mvn', '-Dstorm_version=%s' % version_string, 'assembly:assembly'])
     os.chdir(old_cwd)
     shutil.copyfile(
-        '../jvmpetrel/target/storm-petrel-%s-SNAPSHOT.jar' % version_number,
+        '../jvmpetrel/target/storm-petrel-%s-SNAPSHOT-jar-with-dependencies.jar' % version_number,
```
Comment (barrywhart) [inline petrel/setup.py]: Please explain the reason for this change.

File: petrel/petrel/cmdline.py (hunk 2 above)
Comment (grant-traynor) [inline petrel/petrel/cmdline.py]: If we don't want jars in the repository, should they not be in the .gitignore? That was the logic I used.

File: petrel/setup.py (hunk 6 above)
Comment (grant-traynor) [inline petrel/setup.py]: Had trouble finding snakeyaml when you `petrel submit`.

File: petrel/setup.py (hunk 5 above)
Comment (barrywhart) [inline petrel/setup.py]: What do you mean by "elsewhere in the repository"?

File: petrel/setup.py (hunk 3 above)
Comment (grant-traynor) [inline petrel/setup.py]: OK. Can fix this. What would you like the version number to be?

File: petrel/setup.py (hunk 5 above)
Comment (grant-traynor) [inline petrel/setup.py]: ./Petrel/petrel/petrel/cmdline.py:    m = re.search('^(Storm )?(\d+\.\d+\.\d+)', version)

File: petrel/petrel/cmdline.py (hunk 2 above)
Comment (barrywhart) [inline petrel/petrel/cmdline.py]: >Because in our environment the storm binary is linked, which creates a reference to /usr.

I don't follow your logic here.

Comment (barrywhart) [inline petrel/petrel/cmdline.py]: >If we don't want jars in the repository, should they not be in the .gitignore? That was the logic I used.

Generated files should almost never go in source control. You can change .gitignore if you want.

File: petrel/setup.py (hunk 3 above)
Comment (barrywhart) [inline petrel/setup.py]: Let's keep it at 0.3 for now.

File: petrel/setup.py (hunk 5 above)
Comment (barrywhart) [inline petrel/setup.py]: Ah, good point. :)

File: petrel/setup.py (hunk 6 above)
Comment (barrywhart) [inline petrel/setup.py]: Who uses snakeyaml, Storm or your topology?

File: petrel/petrel/cmdline.py (hunk 2 above)
Comment (grant-traynor) [inline petrel/petrel/cmdline.py]: OK. I'll change .gitignore. I have the source code for petrel checked out in /opt/petrel, but also installed it via `python config.py install`, hence it links to `/usr/local/bin`, but the script that build topologies seems to want the /opt/petrel (checked out code)? Might have misunderstood this.

File: petrel/setup.py (hunk 3 above)
Comment (grant-traynor) [inline petrel/setup.py]: OK.

File: petrel/setup.py (hunk 5 above)
Comment (grant-traynor) [inline petrel/setup.py]: Happy to make it whatever you want. Too lazy to invent it myself :-).

File: petrel/setup.py (hunk 6 above)
Comment (grant-traynor) [inline petrel/setup.py]: Not sure. It gets put into the jvmpetrel .jar, which is build outside the topology? So I'm assuming it's Petrel that wants it?

**Comment by barrywhart** [comment]
I updated Petrel today to work with Storm 0.10.0. Many of these changes were useful. Thanks!
